modules/e2ee/Worker.js

```diff
@@ -247,61 +273,69 @@ class Context {
-     * 6) Append the plaintext to the decrypted frame.
-     * 7) Enqueue the decrypted frame for decoding.
      */
     decodeFunction(encodedFrame, controller) {
         const data = new Uint8Array(encodedFrame.data);
-        const keyIndex = data[encodedFrame.data.byteLength - 1];
+        const keyIndex = data[encodedFrame.data.byteLength - 1] & 0x7; // eslint-disable-line no-bitwise
 
         if (this._cryptoKeyRing[keyIndex]) {
-            const iv = new Uint8Array(encodedFrame.data, encodedFrame.data.byteLength - ivLength - 1, ivLength);
-            const cipherTextStart = unencryptedBytes[encodedFrame.type];
-            const cipherTextLength = encodedFrame.data.byteLength - (unencryptedBytes[encodedFrame.type]
-                + ivLength + 1);
-
-            return crypto.subtle.decrypt({
-                name: 'AES-GCM',
-                iv,
-                additionalData: new Uint8Array(encodedFrame.data, 0, unencryptedBytes[encodedFrame.type])
-            }, this._cryptoKeyRing[keyIndex], new Uint8Array(encodedFrame.data, cipherTextStart, cipherTextLength))
-            .then(plainText => {
-                const newData = new ArrayBuffer(unencryptedBytes[encodedFrame.type] + plainText.byteLength);
-                const newUint8 = new Uint8Array(newData);
+            // eslint-disable-next-line no-bitwise
+            const counterLength = 1 + ((data[encodedFrame.data.byteLength - 1] >> 4) & 0x7);
+            const frameHeader = new Uint8Array(encodedFrame.data, 0, unencryptedBytes[encodedFrame.type]);
 
-                newUint8.set(new Uint8Array(encodedFrame.data, 0, unencryptedBytes[encodedFrame.type]));
-                newUint8.set(new Uint8Array(plainText), unencryptedBytes[encodedFrame.type]);
+            // Extract the signature.
+            const signatureOffset = encodedFrame.data.byteLength - (digestLength[encodedFrame.type]
+                + counterLength + 1);
+            const signature = encodedFrame.data.slice(signatureOffset, signatureOffset
+                + digestLength[encodedFrame.type]);
 
-                encodedFrame.data = newData;
+            // Set signature bytes to 0.
+            const zeros = new Uint8Array(digestLength[encodedFrame.type]);
 
-                return controller.enqueue(encodedFrame);
-            }, e => {
-                console.error(e);
+            data.set(zeros, encodedFrame.data.byteLength - (digestLength[encodedFrame.type] + counterLength + 1));
+
+            return crypto.subtle.sign(signatureOptions, this._cryptoKeyRing[keyIndex].authenticationKey,
+                encodedFrame.data).then(calculatedSignature => {
+                // Do truncated hash comparison.
+                if (!isArrayEqual(signature, calculatedSignature.slice(0, digestLength[encodedFrame.type]))) {
+                    console.error('signature mismatch', new Uint8Array(signature), new Uint8Array(calculatedSignature,
```

Comment:
push that out of the worker somehow

Reply:
Can you put a TODO comment here? We will tackle it right after this lands.